README.md
# less-tedious

Helper package for the node.js tedious SQL Server library, and accompanying mssql client library.

This package provides the following...

- easy to use `query` and `modify` methods that simplifies common SQL calls
- automatic connection management and pool initialisation
- verbose debugging info, enabled by `winston`
- retry logic which tolerates timeouts and other custom conditions of your choosing


## Async Retry mechanism

The async retry handler was created to support the retry of database calls under certain conditions.
For example, when all database connections are saturated, you may want to wait 5 seconds before re-attempting.

### Example

```javascript
const retry = require('./retry-async')
// actual call to database wrapped in an async function... 
const callDatabase = async () => {
  const data = await sql.query('SELECT * FROM [SomeTable]')
  return data
}
// we only want to retry timeout errors...
const onlyRetryTimeoutsPredicate = (error) => {
  return error.message.contains('timeout')
}

const retryPolicy = {
  attempts: 3,
  pauseTimeMs: 5000,
  pauseMultiplier: 1.5
}

const attemptCount = 3
try {
  const data = await retry(callDatabase, retryPolicy, onlyRetryTimeoutsPredicate)
  console.log(`we got some data:${data}`)
} catch (error) {
  console.error(`attempted to query database ${retryPolicy.attempts} times, but all calls were unsuccessful`)
}
```


CHANGELOG.md
# CHANGELOG

# 1.1.0

* First working version, some features consumed by [MTC](https://github.com/DFEAGILEDEVOPS/MTC)

# 1.1.1

* reduce logfile noise

# 2.0.0

* implement mssql connection pooling
* implement retry policy for sql server 'too many connections' errors

# 2.0.1

* re-introduce sqlService.initialise() to minimise client disruption


spec/integration/retry.spec.js
'use strict'
/* global describe beforeAll it expect fail xit spyOn */

const retry = require('../../retry')
const sqlConfig = require('../../example-config')
const sql = require('../../index')

describe('retry:integration', () => {
  beforeAll(async () => {
    await sql.initPool(sqlConfig)
    await sql.updateDataTypeCache()
  })

  it('should be transparent when handled method works correctly', async () => {
    const handledFn = async () => {
      const data = await sql.query('SELECT * FROM Settings')
      return data
    }
    let settingsRows = await retry(3, handledFn)
    expect(settingsRows).toBeDefined()
    expect(settingsRows.length).toBe(1)
    expect(settingsRows[0].loadingTimeLimit).toBe(3)
  })

  it('should retry 3 times when handled method fails', async () => {
    const handledFn = async () => {
      const data = await sql.query('SELECT * FROM TableThatDoesNotExist')
      return data
    }
    const spy = jasmine.createSpy().and.callFake(handledFn)
    try {
      await retry(3, spy)
      fail('error should have been thrown')
    } catch (error) {
      expect(error.message).toContain(`Invalid object name 'TableThatDoesNotExist'`)
    }
    expect(spy).toHaveBeenCalledTimes(3)
  })

  it('it does not retry if predicate condition is not met', async () => {
    const handledFn = async () => {
      const data = await sql.query('SELECT * FROM TableThatDoesNotExist')
      return data
    }
    const predicate = (error) => {
      console.log(`error in predicate is:${error}`)
      // will always fail
      return error.message === 'sql server is busy'
    }
    const spy = jasmine.createSpy().and.callFake(handledFn)
    try {
      await retry(3, spy, predicate)
      fail('error should have been thrown')
    } catch (error) {
      expect(error.message).toContain(`Invalid object name 'TableThatDoesNotExist'`)
    }
    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('it retries specified number of times when predicate met', async () => {
    const handledFn = async () => {
      const data = await sql.query('SELECT * FROM TableThatDoesNotExist')
      return data
    }
    const predicate = (error) => {
      return error.message.startsWith(`Invalid object name 'TableThatDoesNotExist'`)
    }
    const spy = jasmine.createSpy().and.callFake(handledFn)
    try {
      await retry(3, spy, predicate)
      fail('error should have been thrown')
    } catch (error) {
      expect(error.message).toContain(`Invalid object name 'TableThatDoesNotExist'`)
    }
    expect(spy).toHaveBeenCalledTimes(3)
  })
})


spec/integration/sql-config.js
'use strict'

module.exports = {
  database: process.env.SQL_DATABASE || 'mtc',
  server: process.env.SQL_SERVER || 'localhost',
  port: process.env.SQL_PORT || 1433,
  user: process.env.SQL_APP_USER || 'mtcAdminUser', // docker default
  password: process.env.SQL_APP_USER_PASSWORD || 'your-chosen*P4ssw0rd_for_dev_env!', // docker default
  pool: {
    min: process.env.SQL_POOL_MIN_COUNT || 0,
    max: process.env.SQL_POOL_MAX_COUNT || 5
  },
  options: {
    appName: process.env.SQL_APP_NAME || 'less-tedious-integration-tests', // docker default
    encrypt: process.env.SQL_ENCRYPT || true
  }
}


spec/integration/retry-async.spec.js
'use strict'
/* global describe beforeAll it expect fail xit spyOn */

const retry = require('../../retry-async')
const sqlConfig = require('../../example-config')
const sql = require('mssql')

const retryConfiguration = {
  attempts: 3,
  pauseTimeMs: 100,
  pauseMultiplier: 1.5
}

describe('retry:integration', () => {
  beforeAll(async () => {
    const config = {
      user: sqlConfig.Application.Username,
      password: sqlConfig.Application.Password,
      server: sqlConfig.Server,
      database: sqlConfig.Database,
      connectionTimeout: sqlConfig.connectionTimeout || 30000,
      requestTimeout: sqlConfig.requestTimeout || 15000,
      pool: {
        max: sqlConfig.Pooling.MaxCount || 5,
        min: sqlConfig.Pooling.MinCount || 0,
        idleTimeoutMillis: sqlConfig.Pooling.IdleTimeout || 30000
      },
      options: {
        encrypt: sqlConfig.Encrypt
      }
    }
    await sql.connect(config)
  })

  it('should be transparent when handled method works correctly', async () => {
    const handledFn = async () => {
      return sql.query('SELECT * FROM Settings')
    }
    let settingsRows = await retry(handledFn, retryConfiguration)
    expect(settingsRows).toBeDefined()
    expect(settingsRows.recordset.length).toBe(1)
    const record = settingsRows.recordset[0]
    expect(record.loadingTimeLimit).toBe(3)
  })

  it('should retry 3 times when handled method fails', async () => {
    const handledFn = async () => {
      return sql.query('SELECT * FROM TableThatDoesNotExist')
    }
    const spy = jasmine.createSpy().and.callFake(handledFn)
    try {
      await retry(spy, retryConfiguration)
      fail('error should have been thrown')
    } catch (error) {
      expect(error.message).toContain(`Invalid object name 'TableThatDoesNotExist'`)
    }
    expect(spy).toHaveBeenCalledTimes(3)
  })

  it('it does not retry if predicate condition is not met', async () => {
    const handledFn = async () => {
      return sql.query('SELECT * FROM TableThatDoesNotExist')
    }
    const predicate = (error) => {
      // will always fail
      return error.message === 'sql server is busy'
    }
    const spy = jasmine.createSpy().and.callFake(handledFn)
    try {
      await retry(spy, retryConfiguration, predicate)
      fail('error should have been thrown')
    } catch (error) {
      expect(error.message).toContain(`Invalid object name 'TableThatDoesNotExist'`)
    }
    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('it retries specified number of times when predicate met', async () => {
    const handledFn = async () => {
      return sql.query('SELECT * FROM TableThatDoesNotExist')
    }
    const predicate = () => {
      return true
    }
    const spyObject = {
      myFunc: handledFn
    }
    spyOn(spyObject, 'myFunc').and.callThrough()
    try {
      await retry(spyObject.myFunc, retryConfiguration, predicate)
      fail('error should have been thrown')
    } catch (error) {
      expect(error.message).toContain(`Invalid object name 'TableThatDoesNotExist'`)
      expect(spyObject.myFunc).toHaveBeenCalledTimes(3)
      return
    }
    fail('error should have been caught')
  })
})


pool.js
'use strict'

const ConnectionPool = require('tedious-connection-pool')
const winston = require('winston')

const sqlPoolService = {}
let mainPool = null
let config

/**
 * Initialise the connection pool.  Called once per application instance
 */
sqlPoolService.init = (sqlConfig) => {
  config = sqlConfig
  // full config details: https://github.com/tediousjs/tedious/blob/master/src/connection.js

  const mainConnectionConfig = {
    appName: config.Sql.Application.Name,
    userName: config.Sql.Application.Username,
    password: config.Sql.Application.Password,
    server: config.Sql.Server,
    options: {
      port: config.Sql.Port,
      database: config.Sql.Database,
      encrypt: true,
      requestTimeout: config.Sql.Timeout
    }
  }

  const mainPoolConfig = {
    min: config.Sql.Pooling.MinCount,
    max: config.Sql.Pooling.MaxCount
  }

  if (mainPool == null) {
    mainPool = new ConnectionPool(mainPoolConfig, mainConnectionConfig)
    mainPool.on('error', function (err) {
      winston.error(err)
    })
  }
}

/**
 * Get a connection from the pool.
 * @return {Promise}
 */
sqlPoolService.getConnection = () => {
  return new Promise((resolve, reject) => {
    if (mainPool === null) {
      sqlPoolService.init(config)
    }
    mainPool.acquire(function (err, connection) {
      if (err) {
        reject(err)
        return
      }
      resolve(connection)
    })
  })
}

/**
 * Disconnect all pool connections
 */
sqlPoolService.drain = () => {
  if (mainPool) {
    mainPool.drain()
  }
}

module.exports = sqlPoolService


retry-async.js
'use strict'

const pause = (duration) => new Promise(res => setTimeout(res, duration), noReject => undefined)
const defaultRetryCondition = () => true
const defaultConfiguration = {
  attempts: 3,
  pauseTimeMs: 5000,
  pauseMultiplier: 1.5
}

/**
 * @param {function} asyncRetryableFunction - the function to execute with retry behaviour
 * @param {object} retryConfiguration - the behaviour of the retry policy.  Default settings are provided
 * @param {function(Error):Boolean} retryCondition - predicate function to determine if the function should be retried.  Defaults to true
 */
const asyncRetryHandler = async (asyncRetryableFunction, retryConfiguration = defaultConfiguration, retryCondition = defaultRetryCondition) => {
  let retryPolicy = {}
  try {
    Object.assign(retryPolicy, retryConfiguration)
    const result = await asyncRetryableFunction()
    return result
  } catch (error) {
    if (retryPolicy.attempts > 1 && retryCondition(error)) {
      await pause(retryPolicy.pauseTimeMs)
      retryPolicy.attempts -= 1
      retryPolicy.pauseTimeMs *= retryConfiguration.pauseMultiplier
      await asyncRetryHandler(asyncRetryableFunction, retryPolicy, retryCondition)
    } else {
      throw error
    }
  }
}

/**
 * @param {number} attempts - the total number of attempts to make at calling the function successfully
 * @param {function} asyncFn - the async function to attempt
 * @param {function(error:Error)} condition - a predicate that allows you to retry only in certain conditions.  Receives the error argument.  For example - this is useful when you only want to retry a database timeout error, but not a syntax error.
 */
const asyncRetryHandlerV1 = async (attempts, asyncFn, condition = defaultRetryCondition) => {
  try {
    const result = await asyncFn()
    return result
  } catch (error) {
    if (attempts > 1 && condition(error)) {
      await pause(defaultConfiguration.pauseTimeMs)
      await asyncRetryHandler(attempts - 1, asyncFn, condition)
    } else {
      throw error
    }
  }
}

module.exports = asyncRetryHandler


date.service.js
'use strict'
const moment = require('moment')
const logger = require('winston')

const gdsFullFormat = 'D MMMM YYYY'
const gdsShortFormat = 'D MMM YYYY'
const UKFormat = 'DD/MM/YYYY'
const yearFormat = 'YYYY'
const reverseFormatNoSeparator = 'YYYYMMDD'
const timeFormatWithSeconds = 'h:mm:ss a'
const dayAndDateFormat = 'dddd, D MMMM'
const dayDateAndYearFormat = 'dddd D MMMM YYYY'
const dateAndTimeFormat = 'D MMMM YYYY h:mma'
const iso8601WithMsPrecisionAndTimeZone = 'YYYY-MM-DDTHH:mm:ss.SSSZ'
const filenameFriendly = 'YYYY-MM-DD-HHmm'

const dateService = {
  formatYear: function (date) {
    return moment(date).format(yearFormat)
  },

  formatFullGdsDate: function (date) {
    return dateService.checkAndFormat(date, gdsFullFormat)
  },

  formatShortGdsDate: function (date) {
    return dateService.checkAndFormat(date, gdsShortFormat)
  },

  formatDayAndDate: function (date) {
    return moment(date).format(dayAndDateFormat)
  },

  formatDayDateAndYear: function (date) {
    return moment(date).format(dayDateAndYearFormat)
  },

  formatDateAndTime: function (date) {
    return moment(date).format(dateAndTimeFormat)
  },

  formatUKDate: function (date) {
    return dateService.checkAndFormat(date, UKFormat)
  },

  reverseFormatNoSeparator: function (date) {
    return dateService.checkAndFormat(date, reverseFormatNoSeparator)
  },

  formatTimeWithSeconds: function (date) {
    return dateService.checkAndFormat(date, timeFormatWithSeconds)
  },

  formatIso8601: function (momentDate) {
    if (!moment.isMoment(momentDate)) {
      throw new Error('Parameter must be of type Moment')
    }
    if (!momentDate.isValid()) {
      throw new Error('Not a valid date')
    }
    return momentDate.format(iso8601WithMsPrecisionAndTimeZone)
  },

  checkAndFormat: function (date, format) {
    if (!(date instanceof Date || moment.isMoment(date))) {
      logger.warn(`Date parameter is not a Date or Moment object: ${date}`)
      return ''
    }
    const m = moment(date)
    if (!m.isValid()) {
      return ''
    }
    return m.format(format)
  },
  /**
   * Format check windows dates.
   * @param dateItem
   * @param keyDay
   * @param keyMonth
   * @param keyYear
   * @returns {Moment}
   */
  formatDateFromRequest: function (dateItem, keyDay, keyMonth, keyYear) {
    return moment.utc(
      '' + dateItem[keyDay] +
      '/' + dateItem[keyMonth] +
      '/' + dateItem[keyYear],
      'D/MM/YYYY')
  },

  /**
   * Format period (start and end dates)
   * @param startDate
   * @param endDate
   * @returns {string} E.g. "1 Nov to 20 Nov 2017" or "1 Dec 2016 to 1 Jan 2017"
   */
  formatCheckPeriod: function (startDate, endDate) {
    let startYear = ' ' + startDate.format('YYYY')
    let endYear = ' ' + endDate.format('YYYY')

    if (startYear === endYear) {
      startYear = ''
    }
    return startDate.format('D MMM') + startYear + ' to ' + endDate.format('D MMM YYYY')
  },

  /**
   * Return a moment object from the a day, month and year. The time component will be zeroed out. Returns null if invalid.
   * @param {number|string} day
   * @param {number|string} month
   * @param {number|string} year
   * @returns {Moment}
   */
  createLocalTimeFromDayMonthYear: function (day, month, year) {
    const paddedDay = (+day).toString().padStart(2, '0')
    const paddedMonth = (+month).toString().padStart(2, '0')
    const data = paddedDay + '/' + paddedMonth + '/' + (+year).toString()
    const date = moment(data, 'DD/MM/YYYY', true)
    if (!date.isValid()) {
      return null
    }
    return date
  },
  /**
   * Return a UTC-mode moment object from the a day, month and year. The time component will be zeroed out. Returns null if invalid.
   * @param {number|string} day
   * @param {number|string} month
   * @param {number|string} year
   * @returns {Moment}
   */
  createUTCFromDayMonthYear: function (day, month, year) {
    const paddedDay = (+day).toString().padStart(2, '0')
    const paddedMonth = (+month).toString().padStart(2, '0')
    const data = paddedDay + '/' + paddedMonth + '/' + (+year).toString()
    const date = moment.utc(data, 'DD/MM/YYYY', true)
    if (!date.isValid()) {
      return null
    }
    return date
  },
  /**
   * returns current UTC date and time as moment
   * @returns {Moment}
   */
  utcNowAsMoment: () => {
    return moment.utc()
  },

  formatFileName: function (date) {
    return moment(date).format(filenameFriendly)
  }
}

module.exports = dateService


index.js
'use strict'

const R = require('ramda')
const moment = require('moment')
let cache = {}
const mssql = require('mssql')
const dateService = require('./date.service')
let pool

/** Utility functions **/

/**
 * Return the Tedious key value given a SQL Server datatype
 * E.g. 'nvarchar' => 'NVarChar'
 * To get the whole Type for a parameter you need the object stored under this key.
 * @param {string} type
 * @return {string | undefined}
 */
const findDataType = (type) => Object.keys(sqlService.TYPES).find(k => {
  if (type.toUpperCase() === k.toUpperCase()) {
    return k
  }
})

/**
 *  Given table and column names returns the cache key for looking up the datatype from the DB server
 *  It will remove square brackets from the table name if present.
 * @param {string} table
 * @param {string} column
 * @return {string}
 */
const cacheKey = (table, column) => table.replace('[', '').replace(']', '') + '_' + column

/**
 * Returns a function that extracts the keys from an object joins them together in a sql fragment
 * @type {Function}
 */
const extractColumns = R.compose(
  R.join(' , '),
  R.keys
)

/**
 * Prefix a string with '@'
 * @param s
 * @return {string}
 */
const paramName = (s) => '@' + s

/**
 * Prefix a string with '@' and its index
 * @param idx
 * @param s
 * @return {string}
 */
const paramNameWithIdx = R.curry((idx, s) => '@' + s + idx)

/**
 * Return a function that takes the keys from an object joins them together into a list
 * of sql parameter identifiers.
 * @type {Function}
 */
const createParamIdentifiers = R.compose(
  R.join(' , '),
  R.map(paramName),
  R.keys
)

/**
 * Return a function that takes the keys from each object in an array and joins them
 * together into a list of sql parameter identifiers with specific IDs.
 * @type {Function}
 */
const createMultipleParamIdentifiers = (data) => data.map((d, idx) => R.compose(
  R.join(' , '),
  R.map(paramNameWithIdx(idx)),
  R.keys
)(d))

/**
 * Return a string for use in a SQL UPDATE statement
 * E.g. 'foo' => 'foo=@foo'
 * @param k
 */
const singleUpdate = (k) => R.join('', [k, '=', paramName(k)])

/**
 * Return a function that generates a list of UPDATE fragments for an entire object's keys (except the `id` key)
 * @type {Function}
 */
const generateSetStatements = R.compose(
  R.join(' , '),
  R.map(singleUpdate),
  R.keys
)

/**
 * Returns a bool indicating the supplied object is a Moment object
 * @param v
 * @return {boolean}
 */
const isMoment = (v) => moment.isMoment(v)

/**
 * Given an object will convert all Moment values to Javascript Date
 * Useful for converting Data during UPDATES and INSERTS
 */
const convertMomentToJsDate = (m) => {
  if (!isMoment(m)) {
    return m
  }
  const iso = dateService.formatIso8601(m)
  return new Date(iso)
}

/**
 * Convert Date to Moment object
 * Useful for converting Data during UPDATES and INSERTS
 */
const convertDateToMoment = (d) => {
  if (!(d instanceof Date)) {
    return d
  }
  return moment(d)
}

/**
 * Return a list of parameters given a table and an object whose keys are column names
 * @param {string} tableName
 * @param {object} data - keys should be col. names
 * @return {Promise<Array>}
 */
async function generateParams (tableName, data) {
  const pairs = R.toPairs(data)
  const params = []
  for (const p of pairs) {
    const [column, value] = p
    const cacheData = await sqlService.getCacheEntryForColumn(tableName, column)
    if (!cacheData) {
      throw new Error(`Column '${column}' not found in table '${tableName}'`)
    }
    const options = {}
    // Construct the options array for params generated used `create()` or `update()`
    if (cacheData.dataType === 'Decimal' || cacheData.dataType === 'Numeric') {
      options.precision = cacheData.precision
      options.scale = cacheData.scale
    } else if (cacheData.maxLength) {
      options.length = cacheData.maxLength
    }

    params.push({
      name: column,
      value,
      type: R.prop(findDataType(cacheData.dataType), sqlService.TYPES),
      options
    })
  }
  return params
}

/** SQL Service **/
const sqlService = {
  // SQL type-mapping adapter.  Add new types as required.
  TYPES: {
    BigInt: mssql.BigInt,
    Bit: mssql.Bit,
    Char: mssql.Char,
    DateTimeOffset: mssql.DateTimeOffset,
    Decimal: mssql.Decimal,
    Float: mssql.Float,
    Int: mssql.Int,
    Numeric: mssql.Numeric,
    NVarChar: mssql.NVarChar,
    Real: mssql.Real,
    SmallInt: mssql.SmallInt,
    UniqueIdentifier: mssql.UniqueIdentifier
  }
}

function validateSqlConfig (config) {
  const ex = (propertyName) => {
    throw new Error(`${propertyName} is required`)
  }
  if (!config.Application.Username) {
    ex('Application.Username')
  }
  if (!config.Application.Password) {
    ex('Application.Password')
  }
  if (!config.Server) {
    ex('Server')
  }
  if (!config.Database) {
    ex('Server')
  }
}

sqlService.initPool = async (sqlConfig) => {
  if (!sqlConfig) {
    throw new Error('sqlConfig is required')
  }
  if (pool) {
    return
  }
  validateSqlConfig(sqlConfig)
  const config = {
    user: sqlConfig.Application.Username,
    password: sqlConfig.Application.Password,
    server: sqlConfig.Server,
    database: sqlConfig.Database,
    connectionTimeout: sqlConfig.connectionTimeout || 30000,
    requestTimeout: sqlConfig.requestTimeout || 15000,
    pool: {
      max: sqlConfig.Pooling.MaxCount || 5,
      min: sqlConfig.Pooling.MinCount || 0,
      idleTimeoutMillis: sqlConfig.Pooling.IdleTimeout || 30000
    },
    options: {
      encrypt: sqlConfig.Encrypt
    }
  }

  pool = new mssql.ConnectionPool(config)
  // TODO emit error
  pool.on('error', () => {})
  return pool.connect()
}

sqlService.drainPool = async () => {
  await pool
  if (!pool) {
    return
  }
  return pool.close()
}

/**
 * Utility service to transform the results before sending to the caller
 * @type {Function}
 */
sqlService.transformResult = function (data) {
  const recordSet = R.prop('recordset', data) // returns [o1, o2,  ...]
  if (!recordSet) {
    return []
  }

  return R.map(R.pipe(
    R.omit(['version']),
    R.map(convertDateToMoment)
  ), recordSet)
}

function addParamsToRequestSimple (params, request) {
  if (params) {
    for (let index = 0; index < params.length; index++) {
      const param = params[index]
      // TODO support other options
      request.input(param.name, param.type, param.value)
    }
  }
}

const retry = require('./retry-async')
const retryConfig = {
  attempts: 3,
  pauseTimeMs: 5000,
  pauseMultiplier: 1.5
}
const connectionLimitReachedErrorCode = 10928

const dbLimitReached = (error) => {
  // https://docs.microsoft.com/en-us/azure/sql-database/sql-database-develop-error-messages
  return error.number === connectionLimitReachedErrorCode // || error.message.indexOf('request limit') !== -1
}

/**
 * Query data from SQL Server via mssql
 * @param {string} sql - The SELECT statement to execute
 * @param {array} params - Array of parameters for SQL statement
 * @return {Promise<*>}
 */
sqlService.query = async (sql, params = []) => {
  await pool

  const query = async () => {
    const request = new mssql.Request(pool)
    addParamsToRequestSimple(params, request)
    const result = await request.query(sql)
    return sqlService.transformResult(result)
  }

  return retry(query, retryConfig, dbLimitReached)
}

/**
 * Add parameters to an SQL request
 * @param {{name, value, type}[]} params - array of parameter objects
 * @param {{}} request -  mssql request
 */
function addParamsToRequest (params, request) {
  if (params) {
    for (let index = 0; index < params.length; index++) {
      let param = params[index]
      param.value = convertMomentToJsDate(param.value)
      if (!param.type) {
        throw new Error('parameter type invalid')
      }
      const options = {}
      if (R.pathEq(['type', 'name'], 'Decimal', param) ||
        R.pathEq(['type', 'name'], 'Numeric', param)) {
        options.precision = param.precision || 28
        options.scale = param.scale || 5
      }
      const opts = param.options ? param.options : options

      if (opts.precision) {
        request.input(param.name, param.type(opts.precision, opts.scale), param.value)
      } else if (opts.length) {
        request.input(param.name, param.type(opts.length), param.value)
      } else {
        request.input(param.name, param.type, param.value)
      }
    }
  }
}

/**
 * Modify data in SQL Server via mssql library.
 * @param {string} sql - The INSERT/UPDATE/DELETE statement to execute
 * @param {array} params - Array of parameters for SQL statement
 * @return {Promise}
 */
sqlService.modify = async (sql, params = []) => {
  await pool

  const modify = async () => {
    const request = new mssql.Request(pool)
    addParamsToRequest(params, request)
    return request.query(sql)
  }

  const returnValue = {}
  const insertIds = []
  let rawResponse

  rawResponse = await retry(modify, retryConfig, dbLimitReached)

  if (rawResponse && rawResponse.recordset) {
    for (let obj of rawResponse.recordset) {
      /* TODO remove this strict column name limitation and
        extract column value regardless of name */
      if (obj && obj.SCOPE_IDENTITY) {
        insertIds.push(obj.SCOPE_IDENTITY)
      }
    }
  }

  if (insertIds.length === 1) {
    returnValue.insertId = R.head(insertIds)
  } else if (insertIds.length > 1) {
    returnValue.insertIds = insertIds
  }
  return returnValue
}

/**
 * Find a row by numeric ID
 * Assumes all table have Int ID datatype
 * @param {string} table
 * @param {number} id
 * @return {Promise<void>}
 */
sqlService.findOneById = async (table, id, schema = '[mtc_admin]') => {
  const paramId = {
    name: 'id',
    type: sqlService.TYPES.Int,
    value: id
  }
  const sql = `
      SELECT *    
      FROM ${schema}.${table}
      WHERE id = @id
    `

  const query = async () => {
    return sqlService.query(sql, [paramId])
  }

  const rows = await retry(query, retryConfig, dbLimitReached)
  return R.head(rows)
}

/**
 * Return the Tedious datatype object required for a particular table and column
 * It's okay if the table name has square brackets around it like '[pupil]'
 * @param {string} table
 * @param {string} column
 * @return {TYPE}
 *
 */
sqlService.getCacheEntryForColumn = async function (table, column) {
  const key = cacheKey(table, column)
  if (R.isEmpty(cache)) {
    // This will cache all data-types once on the first sql request
    await sqlService.updateDataTypeCache()
  }
  if (!cache.hasOwnProperty(key)) {
    return undefined
  }
  const cacheData = cache[key]
  return cacheData
}

/**
 * Provide the INSERT statement for passing to modify and parameters given a key/value object
 * @param {string} table
 * @param {object} data
 * @return {{sql: string, params}}
 */
sqlService.generateInsertStatement = async (table, data, schema = '[mtc_admin]') => {
  const params = await generateParams(table, data)
  const sql = `
  INSERT INTO ${schema}.${table} ( ${extractColumns(data)} ) VALUES ( ${createParamIdentifiers(data)} );
  SELECT SCOPE_IDENTITY() AS [SCOPE_IDENTITY]`
  return {
    sql,
    params,
    outputParams: { SCOPE_IDENTITY: sqlService.TYPES.Int }
  }
}

/**
 * Provide the INSERT statements for passing to modify and parameters an array
 * @param {string} table
 * @param {array} data
 * @return {{sql: string, params}}
 */
sqlService.generateMultipleInsertStatements = async (table, data, schema = '[mtc_admin]') => {
  if (!Array.isArray(data)) throw new Error('Insert data is not an array')
  const paramsWithTypes = await generateParams(table, R.head(data))
  const headers = extractColumns(R.head(data))
  const values = createMultipleParamIdentifiers(data).join('), (')
  let params = []
  data.forEach((datum, idx) => {
    params.push(
      R.map((key) => {
        const sameParamWithType = paramsWithTypes.find(({ name }) => name === key)
        return {
          ...sameParamWithType,
          name: `${key}${idx}`,
          value: (sameParamWithType.type.type === 'DATETIMEOFFSETN' ? moment(datum[key]) : datum[key])
        }
      }, R.keys(datum))
    )
  })
  params = R.flatten(params)
  const sql = `
  INSERT INTO ${schema}.${table} ( ${headers} ) VALUES ( ${values} );
  SELECT SCOPE_IDENTITY()`
  return {
    sql,
    params
  }
}

/**
 * Utility function for internal sqlService use.  Generate the SQL UPDATE statement and list of parameters
 * given the table name and the object containing key/values to be updated.
 * @param table
 * @param data
 * @return {Promise<{sql, params: Array}>}
 */
sqlService.generateUpdateStatement = async (table, data, schema = '[mtc_admin]') => {
  const params = await generateParams(table, data)
  const sql = R.join(' ', [
    `UPDATE ${schema}.${table}`,
    'SET',
    generateSetStatements(R.omit(['id'], data)),
    'WHERE id=@id'
  ])
  return {
    sql,
    params
  }
}

/**
 * Create a new record
 * @param {string} tableName
 * @param {object} data
 * @return {Promise} - returns the number of rows modified (e.g. 1)
 */
sqlService.create = async (tableName, data) => {
  const preparedData = convertMomentToJsDate(data)
  const {
    sql,
    params,
    outputParams
  } = await sqlService.generateInsertStatement(tableName, preparedData)

  const create = async () => {
    return sqlService.modify(sql, params, outputParams)
  }
  return retry(create, retryConfig, dbLimitReached)
}

/**
 * Fetch the data-types for each column in the schema.  Populates the cache.
 * @return {Promise<void>}
 */
sqlService.updateDataTypeCache = async function () {
  const sql =
    `SELECT  
      TABLE_NAME, 
      COLUMN_NAME, 
      DATA_TYPE, 
      NUMERIC_PRECISION, 
      NUMERIC_SCALE, 
      CHARACTER_MAXIMUM_LENGTH
    FROM INFORMATION_SCHEMA.COLUMNS
    WHERE TABLE_SCHEMA = @schema
    `
  const paramSchema = {
    name: 'schema',
    value: 'mtc_admin',
    type: sqlService.TYPES.NVarChar
  }
  // delete any existing cache
  cache = {}
  const rows = await sqlService.query(sql, [paramSchema])
  rows.forEach((row) => {
    const key = cacheKey(row.TABLE_NAME, row.COLUMN_NAME)
    // add the datatype to the cache
    cache[key] = {
      dataType: findDataType(row.DATA_TYPE),
      precision: row.NUMERIC_PRECISION,
      scale: row.NUMERIC_SCALE,
      maxLength: row.CHARACTER_MAXIMUM_LENGTH && row.CHARACTER_MAXIMUM_LENGTH > 0 ? row.CHARACTER_MAXIMUM_LENGTH : undefined
    }
  })
}

/**
 * Call SQL Update on the table given an object whose keys are the columns to be updated and whose keys are the new
 * values. You *MUST* pass the `id` field for the WHERE clause.
 * Returns { rowsModified: n } the number of rows modified.
 * @param tableName
 * @param data
 * @return {Promise<*>}
 */
sqlService.update = async function (tableName, data) {
  if (!data.id) {
    throw new Error('`id` is required')
  }
  // Convert any moment objects to JS Date objects as that's required by Tedious
  const preparedData = convertMomentToJsDate(data)
  const {
    sql,
    params
  } = await sqlService.generateUpdateStatement(tableName, preparedData)

  const update = async () => {
    return sqlService.modify(sql, params)
  }

  return retry(update, retryConfig, dbLimitReached)
}

/**
 * Helper function useful for constructing parameterised WHERE clauses
 * @param {Array} ary
 * @param {sqlService.TYPE} type
 * @return {Promise<{params: Array, paramIdentifiers: Array}>}
 */
sqlService.buildParameterList = (ary, type) => {
  const params = []
  const paramIdentifiers = []
  for (let i = 0; i < ary.length; i++) {
    params.push({
      name: `p${i}`,
      type,
      value: ary[i]
    })
    paramIdentifiers.push(`@p${i}`)
  }
  return {
    params,
    paramIdentifiers
  }
}

sqlService.modifyWithTransaction = async (sqlStatements, params) => {
  const wrappedSQL = `
  BEGIN TRY
  BEGIN TRANSACTION
    ${sqlStatements}
 COMMIT TRANSACTION
END TRY

BEGIN CATCH
  IF (@@TRANCOUNT > 0)
   BEGIN
      ROLLBACK TRANSACTION
      PRINT 'Error detected, all changes reversed'
   END
  DECLARE @ErrorMessage NVARCHAR(4000);
    DECLARE @ErrorSeverity INT;
    DECLARE @ErrorState INT;

    SELECT @ErrorMessage = ERROR_MESSAGE(),
           @ErrorSeverity = ERROR_SEVERITY(),
           @ErrorState = ERROR_STATE();

    -- Use RAISERROR inside the CATCH block to return
    -- error information about the original error that
    -- caused execution to jump to the CATCH block.
    RAISERROR (@ErrorMessage, -- Message text.
               @ErrorSeverity, -- Severity.
               @ErrorState -- State.
               );
END CATCH
  `
  const modify = async () => sqlService.modify(wrappedSQL, params)
  return retry(modify, retryConfig, dbLimitReached)
}

/**
 * Initialise the sql service and set up the connection pool
 * @param config
 */
sqlService.initialise = async (config) => {
  await sqlService.initPool(config)
}

module.exports = sqlService
